File: pocketr-java.api/src/main/java/com/decrux/pocketr/api/services/user_registration/RegisterUserImpl.java
package com.decrux.pocketr.api.services.user_registration;

import com.decrux.pocketr.api.entities.db.auth.User;
import com.decrux.pocketr.api.entities.db.auth.UserRole;
import com.decrux.pocketr.api.entities.dtos.RegisterUserDto;
import com.decrux.pocketr.api.repositories.UserRepository;
import java.util.ArrayList;
import java.util.List;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class RegisterUserImpl implements RegisterUser {

    private static final String ROLE_USER = "USER";

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public RegisterUserImpl(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    @Override
    @Transactional
    public void registerUser(RegisterUserDto userDto) {
        String email = userDto.getEmail().trim();

        UserRole role = new UserRole();
        role.setRole(ROLE_USER);

        User user = new User();
        user.setPassword(requireNotNull(passwordEncoder.encode(userDto.getPassword().trim()), "Password encoder returned null"));
        user.setEmail(email);
        user.setFirstName(userDto.getFirstName());
        user.setLastName(userDto.getLastName());

        List<UserRole> roles = new ArrayList<>();
        roles.add(role);
        user.setRoles(roles);

        userRepository.saveAndFlush(user);
    }

    private static <T> T requireNotNull(T value, String message) {
        if (value == null) {
            throw new IllegalArgumentException(message);
        }
        return value;
    }
}

File: pocketr-api/src/main/kotlin/com/decrux/pocketr_api/services/user_registration/RegisterUserImpl.kt
package com.decrux.pocketr_api.services.user_registration

import com.decrux.pocketr_api.entities.db.auth.User
import com.decrux.pocketr_api.entities.db.auth.UserRole
import com.decrux.pocketr_api.entities.dtos.RegisterUserDto
import com.decrux.pocketr_api.repositories.UserRepository
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class RegisterUserImpl(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
) : RegisterUser {
    @Transactional
    override fun registerUser(userDto: RegisterUserDto) {
        val email = userDto.email.trim()

        val user =
            User(
                password =
                    requireNotNull(passwordEncoder.encode(userDto.password.trim())) {
                        "Password encoder returned null"
                    },
                email = email,
                firstName = userDto.firstName,
                lastName = userDto.lastName,
                roles = mutableListOf(UserRole(role = ROLE_USER)),
            )
        userRepository.saveAndFlush(user)
    }

    private companion object {
        const val ROLE_USER = "USER"
    }
}
